src/server/learnocaml_server.ml

```diff
@@ -176,27 +176,55 @@ module Request_handler = struct
     | Response ({contents; _} as r) -> Response {r with contents = f contents}
     | (Cached _ | Status _) as r -> r
 
+  let alphanum = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
+  let alphanum_len = String.length alphanum
+
+  let nonce_req : (Ipaddr.t option, string) Hashtbl.t = Hashtbl.create 533
+
```

Comment:
Hmm, why do you need an `option` in the type of key?

Reply:
I am not sure we always have access to an IP, (I use the result of `ipaddr_of_endp`, just below, and I am not sure what to do if I don't have a TCP connection. The other possibilities are here: https://github.com/mirage/ocaml-conduit/blob/master/lib/conduit.mli#L52 ).

Reply:
Couldn't you use the `endp` as a key? (After filtering the `Unknown case, of course as it should result in a user error.)

Reply:
The problem is that `endp` contains also the port for TCP, and the same client can send request from the same ip, but its output port is not guaranteed to be same (I hit this issue, and I switched to use only the ip).

Reply:
I see. In any case, the `None` key can be reused several times by two clients with no IP. That's a security problem, probably.

Couldn't you write a function of type `endp -> string` that would improve this situation (string being used as type for keys)?

Reply:
You are right, I have made a `string_of_endp`